src/v/cluster/rm_stm.cc

```diff
@@ -783,6 +792,120 @@ ss::future<bool> rm_stm::sync(model::timeout_clock::duration timeout) {
     co_return ready;
 }
 
+void rm_stm::track_tx(
+  model::producer_identity pid,
+  std::chrono::milliseconds transaction_timeout_ms) {
+    if (_gate.is_closed()) {
+        return;
+    }
+    _mem_state.expiration[pid] = expiration_info{
+      .timeout = transaction_timeout_ms, .last_update = clock_type::now()};
+    auto deadline = _mem_state.expiration[pid].deadline();
+    if (
+      (auto_abort_timer.armed() && auto_abort_timer.get_timeout() > deadline)
+      || !auto_abort_timer.armed()) {
+        auto_abort_timer.cancel();
+        auto_abort_timer.arm(deadline);
+    }
+}
+
+void rm_stm::abort_old_txes() {
+    (void)ss::with_gate(_gate, [this] {
+        std::vector<model::producer_identity> pids;
+        for (auto& [k, _] : _mem_state.estimated) {
+            pids.push_back(k);
+        }
+        for (auto& [k, _] : _mem_state.tx_start) {
+            pids.push_back(k);
+        }
+        for (auto& [k, _] : _log_state.ongoing_map) {
+            pids.push_back(k);
+        }
+        std::vector<model::producer_identity> expired;
+        for (auto pid : pids) {
+            if (_log_state.prepared.contains(pid)) {
+                continue;
+            }
+            if (_mem_state.preparing.contains(pid)) {
+                continue;
+            }
+            if (_mem_state.expected.contains(pid)) {
+                auto expiration_it = _mem_state.expiration.find(pid);
+                if (expiration_it != _mem_state.expiration.end()) {
+                    if (expiration_it->second.deadline() > clock_type::now()) {
+                        continue;
+                    }
+                }
+            }
+            expired.push_back(pid);
+        }
+        return abort_old_txes(expired);
+    });
+}
+
+ss::future<>
+rm_stm::abort_old_txes(std::vector<model::producer_identity> pids) {
+    for (auto pid : pids) {
+        co_await try_abort_old_tx(pid);
+    }
+
+    std::optional<time_point_type> earliest_deadline;
+    for (auto& [_, expiration] : _mem_state.expiration) {
+        auto candidate = expiration.deadline();
+        if (earliest_deadline) {
+            earliest_deadline = std::min(earliest_deadline.value(), candidate);
+        } else {
+            earliest_deadline = candidate;
+        }
+    }
+
+    if (earliest_deadline) {
+        auto_abort_timer.arm(earliest_deadline.value());
```

Comment:
the `arm` method hard asserts that the timer is not armed. since this timer callback runs in the background, `track_tx` may run asynchronously and set the timer before this is evaluated. is that a race that should be handled / am I missing how it is handled?

Reply:
good catch!